### ISSUE TITLE
map-builtin-not-iterating reported with reduce()

### ISSUE THREAD

**Issue by DylanYoung** [PyCQA/pylint#3051]
### Steps to reproduce
1.  Create reduce call with a `map` iterable
2. Observe pylint complain

### Current behavior
1. W1636(map-builtin-not-iterating) is reported

### Expected behavior
1. No error is reported (reduce is obviously an iterating context)

### pylint --version output
pylint 1.9.5, 
astroid 1.6.6
Python 2.7.16 (default, Apr 26 2019, 16:26:18) 
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]

**Comment by DylanYoung** [comment]
Note: it looks like a similar bug was fixed here: https://github.com/PyCQA/pylint/issues/2731

But by the looks of it, I suspect it didn't touch either the Py2 case or the six case of using reduce (it explicitly check fro an 'itertools' root), but I could be mistaken.

**Comment by PCManticore** [comment]
Thanks for the report. Yeah, we'll need a separate solution for `reduce`.

**Comment by gyermolenko** [comment]
@DylanYoung 
Could you please provide an example? I tried this one
```python2
print reduce(lambda x, y: x + y, map(str.upper, "abc"))
```
but didn't get your warning in results

**Comment by DylanYoung** [comment]
@gyermolenko Did you do a `from six.moves import reduce`

**Comment by DylanYoung** [comment]
@PCManticore Moreover, I am still getting an error with the `itertools.chain`:

`from itertools import chain; chain(d1.items(), d2.items())`

**Comment by DylanYoung** [comment]
@PCManticore Also, a question, is there a config option to set other iterating contexts?  This could be a good temporary work-around and is probably necessary to fully support this type of error code (dependent on the context of user code, which could be a c-extension that pylint knows nothing about).

**Comment by PCManticore** [comment]
@DylanYoung Good question, there isn't, but it might be worthwhile to introduce a new option that accepts a list of iterable consuming functions.

**Comment by gyermolenko** [comment]
To summarize:

Warning in question (`(map-builtin-not-iterating)`) 
- appears only with `--py3k` option
- reproducible both for pylint 1.9.5 / python 2.7.16 with python2 example
```python
print reduce(lambda x, y: x + y, map(str.upper, "abc"))
```
and for pylint 2.3.1 / python 3.7.4 with python3 example
```python3
from functools import reduce
print(reduce(lambda x, y: x + y, map(str.upper, "abc")))
```

- seems like "(using six or not)" can be removed from the title


====================================================

As for the second issue, with example
```python
from itertools import chain
d1, d2 = {}, {}
chain(d1.items(), d2.items())
```
warning `(dict-items-not-iterating)` appears only for python2 and only with `--py3k`

**Comment by Pierre-Sassoulas** [comment]
Won't do, we removed the python3 porting mode in 2.11.
